include/counter.h

```diff
@@ -27,30 +27,71 @@
 extern "C" {
 #endif
 
-typedef void (*counter_callback_t)(struct device *dev, void *user_data);
+/** @brief Single shot callback
+ *
+ * @param dev       Pointer to the device structure for the driver instance.
+ * @param id        Single shot request id.
+ * @param counter   Counter value that triggered the callback.
+ * @param user_data User data.
+ */
+typedef void (*counter_single_shot_callback_t)(struct device *dev,
+					       u32_t id,
+					       u32_t counter,
+					       void *user_data);
+
+struct counter_single_shot {
+	/** Index of the request. Maximal index is implementation dependent.*/
+	u8_t id;
+	/** If true ticks are treated as absolute value, else it is relative to
+	 * the counter reading performed during the call.
+	 */
+	bool abs;
+	u32_t ticks;
+	counter_single_shot_callback_t callback;
+	void *user_data;	
+};
+
+/* backward compatibility */
+#define counter_set_alarm counter_set_overflow
+
+typedef void (*counter_callback_t)(struct device *dev,
+					    void *user_data);
 
 typedef int (*counter_api_start)(struct device *dev);
 typedef int (*counter_api_stop)(struct device *dev);
 typedef u32_t (*counter_api_read)(struct device *dev);
-typedef int (*counter_api_set_alarm)(struct device *dev,
-				     counter_callback_t callback,
-				     u32_t count, void *user_data);
+typedef int (*counter_api_set_single_shot)(struct device *dev,
+					   struct counter_single_shot *params);
+
+typedef int (*counter_api_set_overflow)(struct device *dev,
+					u32_t ticks,
+					counter_callback_t callback,
+					void *user_data);
+
 typedef u32_t (*counter_api_get_pending_int)(struct device *dev);
+typedef int (*counter_api_ms_to_ticks)(struct device *dev,
+				       u32_t ms,
+				       u32_t *ticks );
+typedef int (*counter_api_us_to_ticks)(struct device *dev,
+				       u32_t us,
+				       u32_t *ticks );
 
 struct counter_driver_api {
 	counter_api_start start;
 	counter_api_stop stop;
 	counter_api_read read;
-	counter_api_set_alarm set_alarm;
+	counter_api_set_single_shot set_single_shot;
+	counter_api_set_overflow set_overflow;
 	counter_api_get_pending_int get_pending_int;
+	counter_api_ms_to_ticks ms_to_ticks;
+	counter_api_us_to_ticks us_to_ticks;
 };
 
 /**
  * @brief Start counter device in free running mode.
  *
- * Start the counter device. If the device is a 'countup' counter, the
- * counter initial value is set to zero. If it is a 'countdown' counter,
- * the initial value is set to the maximum value supported by the device.
+ * Start the counter device. The counter initial value is set to zero. Counter
+ * wraps at 0xffffffff.
```

Comment:
What about about 16, 24 and 64-bit counters?
What about counting down counters?

A more general questions: 
- Are we going to create low-level API (exposing HW directly) or a higher level thing which implements a generic timer/counter?
- Are we going to use such counter as a tick generator?

Reply:
This API is meant to be generic. When it comes to hw specific stuff (such as clock source, etc...) that's up to the driver.

16, 24 bits counter -> that fits in 32.
Is there 64 bits counter on 32 bits arch? (could be unrelated, thus why I am asking)

counting down/up does not matter much, besides driver implementation that is. I guess.

Reply:
yes, my plan was to hide 16,24 bits capability inside the driver. If hw does not support it then it must emulate it. Otherwise we loose abstraction. And if user is aware of hw limitation set_overflow can set top value within the limiation and emulation will not be used.

64 bits - i don't know. Any generic interface cuts extremes. If 64bit general purpose timer is present and needed it will have to go through custom driver. If zephyr supports more and more devices that have 64 bit timers i would prefer that API is extended with dedicated 64 bit functions.

Reply:
Counting up/down matters if you would like to specify absolute value. If current is tick count is 100 and you specify 200 as a next alarm, you will get different results when your counter counts up and down :)

Reply:
I agree that 32bit emulation is generic and could be done by layer above the driver to make the driver simpler. Probably same with up/down. Configuration of counting source (clock vs other events) is hw dependent and should be done outside (e.g. kconfig option specific to driver implementation or external call to lower layer driver).

In order to make counter.h closer to hw and allow building generic timer on that it would need following changes:
- renaming counter_set_overflow to something like counter_set_top() which would set overflow/reload value depending on direction.
- adding function counter_get_max_top() which would return capacity of the driver.
- adding counter_get_max_alarm() which would return maximal value that can be user for alarm (duration and period)
- adding counter_get_dir() - up or down. I assume that direction would be fixed and configured at compile time (if hw supports up and down)

Reply:
I also see now a need to add ticks_to_us(). It is needed to put counter_read() value on scale.